Make computation of directionality account for Shadow DOM.

dbaron opened the pull request:
This specifies (with some additional detail) the proposal in https://github.com/whatwg/html/issues/3699#issuecomment-951423468 . This changes three things:
 * the inheritance of directionality
 * the inheritance of language
 * the computation of dir=auto to account for Shadow DOM.

This builds on the work in #9452 and #9554 to refactor this section, and builds on work by Brian Kardell, Eric Meyer, and others in #7424 and in #9166 and work by fantasai, rniwa, smaug, MyIdShin, Brian Kardell, and others in #3699.

Closes #9166.

Fixes #3699.

<!--
Thank you for contributing to the HTML Standard! Please describe the change you are making and complete the checklist below if your change is not editorial.
-->

Some aspects of the PR template are copied from #9166.

- [X] At least two implementers are interested (and none opposed):
   * Apple
   * Google
   * Mozilla
   * Igalia
- [x] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * tests pending landing in web-platform-tests/wpt#29820
   * tests pending landing in https://github.com/web-platform-tests/wpt/pull/42364
   * tests already landed recently in:
     * web-platform-tests/wpt/pull/41584
     * web-platform-tests/wpt/pull/41620
     * web-platform-tests/wpt/pull/42030
   * other existing tests in `html/dom/elements/global-attributes/dir*` and `css/selectors/dir*`
   * test plan in web-platform-tests/wpt#25569
- [X] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chromium: https://bugs.chromium.org/p/chromium/issues/detail?id=576815
   * Gecko: https://bugzilla.mozilla.org/show_bug.cgi?id=1722762
   * WebKit: https://bugs.webkit.org/show_bug.cgi?id=64861
- [X] [MDN issue](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) is filed: mdn/mdn#455

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/9796/dom.html" title="Last updated on Oct 6, 2023, 1:37 PM UTC (e0dfa9f)">/dom.html</a>  ( <a href="https://whatpr.org/html/9796/e4ce01e...e0dfa9f/dom.html" title="Last updated on Oct 6, 2023, 1:37 PM UTC (e0dfa9f)">diff</a> )
<a href="https://whatpr.org/html/9796/rendering.html" title="Last updated on Oct 6, 2023, 1:37 PM UTC (e0dfa9f)">/rendering.html</a>  ( <a href="https://whatpr.org/html/9796/e4ce01e...e0dfa9f/rendering.html" title="Last updated on Oct 6, 2023, 1:37 PM UTC (e0dfa9f)">diff</a> )

dbaron commented:
This PR is a revised version of #9166 based on the refactoring mentioned above.  I'm sure there are some edge cases that come out differently.  Another thing that I intended to be different is what is described in https://github.com/whatwg/html/issues/3699#issuecomment-1687073853 .

One other thing that I noticed while writing this is that the rules for processing text directionality are slightly different for an `input` or `textarea` value versus text content -- when processing a value the result is `ltr` when there is a non-empty string that contains no characters of classes L, AL, or R, but falls back to the parent only when the string is empty.  But for processing text nodes, the computation falls back to the parent whenever there are no characters of classes L, AL, or R.

I'm hoping at least some of @fantasai @rniwa @bkardell and @annevk are able to review.  (I can't assign reviewers, it seems.)

dbaron commented:
One other review note is that the way I ended up needing to use the concept of "auto directionality", it ended up easier to redefine it so that it can be null if no directionality is computed rather than including the fallback to parent directionality, and the user of the "auto directionality" is now responsible for replacing null with the parent directionality (or, in some new cases, continuing on to the next element or similar).

dbaron commented:
Also adding @smaug---- for review and feedback on implementer interest from Mozilla.

annevk commented on source:
```diff
@@ -13274,8 +13274,18 @@ Transport Protocol">HTTP&lt;/abbr> today.&lt;/p></code></pre> <!-- DO NOT REWRAP
+
+  <dl class="switch">
+   <dt>If the element's parent is a <span>shadow root</span></dt>
+   <dt>If the element itself is a <code>slot</code> element that has a <span
+   data-x="concept-element-shadow-root">shadow root</span> as its root</dt>
```
```suggestion
   <dt>If the element is a <code>slot</code> element whose <span
   data-x="concept-tree-root">root</span> is a <span>shadow root</span></dt>
```
concept-element-shadow-root is a member of an element.

annevk commented on source:
```diff
@@ -13274,8 +13274,18 @@ Transport Protocol">HTTP&lt;/abbr> today.&lt;/p></code></pre> <!-- DO NOT REWRAP
+   <dt>If the element's parent is a <span>shadow root</span></dt>
+   <dt>If the element itself is a <code>slot</code> element that has a <span
+   data-x="concept-element-shadow-root">shadow root</span> as its root</dt>
+   <dd>the same as the language of its <span>shadow root</span>'s <span
```
```suggestion
   <dd>the same as the language of that <span>shadow root</span>'s <span
```
("its" is very easy to confuse with concept-element-shadow-root, which we don't want here.)

annevk commented on source:
```diff
@@ -13274,8 +13274,18 @@ Transport Protocol">HTTP&lt;/abbr> today.&lt;/p></code></pre> <!-- DO NOT REWRAP
+   <dt>If the element itself is a <code>slot</code> element that has a <span
+   data-x="concept-element-shadow-root">shadow root</span> as its root</dt>
+   <dd>the same as the language of its <span>shadow root</span>'s <span
+   data-x="dom-hyperlink-host">host</span> element, if any.</dd>
```
```suggestion
   data-x="dom-hyperlink-host">host</span>, if any.</dd>
```

annevk commented on source:
```diff
@@ -13274,8 +13274,18 @@ Transport Protocol">HTTP&lt;/abbr> today.&lt;/p></code></pre> <!-- DO NOT REWRAP
 
-  <p>If these attributes are omitted from an element, then the language of this element is the same
-  as the language of its parent element, if any.</p>
+  <p>If these attributes are omitted from an element, then the language of this element is:</p>
```
Looking at this section as a whole it seems that we actually need to change the computation of "language" which is defined below in the implementer section.

annevk commented on source:
```diff
@@ -13586,47 +13606,99 @@ Transport Protocol">HTTP&lt;/abbr> today.&lt;/p></code></pre> <!-- DO NOT REWRAP
 
-   <li><p>Let <var>codePoint</var> be null.</p></li>
+   <li>
+    <p>If <var>element</var> is a <code>slot</code> element that is in a <span>shadow
+    tree</span> whose <span>assigned nodes</span> are not empty:</p>
```
Instead of "in a shadow tree" I think using the "root is a shadow root" check is better. It's more explicit.

annevk commented on source:
```diff
@@ -13586,47 +13606,99 @@ Transport Protocol">HTTP&lt;/abbr> today.&lt;/p></code></pre> <!-- DO NOT REWRAP
+
+    <ol>
+     <li>
+      <p>For each node <var>child</var> in <var>element</var>'s <span>assigned nodes</span>:</p>
```
of

annevk replied:
And use "list iterate" here too for for each?

annevk commented on source:
```diff
@@ -13586,47 +13606,99 @@ Transport Protocol">HTTP&lt;/abbr> today.&lt;/p></code></pre> <!-- DO NOT REWRAP
        <li>a <code>script</code> element</li>
        <li>a <code>style</code> element</li>
        <li>a <code>textarea</code> element</li>
+       <li>a <code>slot</code> element that is in a <span>shadow tree</span></li>
```
whose root is a shadow root (I see this applies a few more times, let's discuss if needed in one comment and settle on a consistent style)

annevk commented on source:
```diff
@@ -13586,47 +13606,99 @@ Transport Protocol">HTTP&lt;/abbr> today.&lt;/p></code></pre> <!-- DO NOT REWRAP
+       <li><p>Let <var>result</var> be the <span>auto directionality</span> of
+       <var>descendant</var>.</p></li>.
+
+       <li><p>If <var>result</var> is null, then set <var>result</var> to '<span
+       data-x="concept-ltr">ltr</span>'.</p></li>
```
Use an early return here?

annevk commented on source:
```diff
@@ -13586,47 +13606,99 @@ Transport Protocol">HTTP&lt;/abbr> today.&lt;/p></code></pre> <!-- DO NOT REWRAP
-   <li><p>If <var>codePoint</var> is null, then return the <span>parent directionality</span> of
-   <var>element</var>.</p></li>
+   <li><p>Return null.</p></li>
+
```
This newline can be removed.

annevk commented on source:
```diff
@@ -13638,11 +13710,19 @@ Transport Protocol">HTTP&lt;/abbr> today.&lt;/p></code></pre> <!-- DO NOT REWRAP
+
+   <li><p>Let <var>parentNode</var> be <var>element</var>'s parent node.</p></li>
+
+   <li><p>If <var>parentNode</var> is a <span>shadow root</span>, then return the directionality of
```
Need to xref "the directionality"

dbaron commented:
Thanks for the review, @annevk.  I think I've addressed all of your comments.  Two notes:
* the only one that was somewhat involved was the first one, regarding fixing the implementer requirements for language calculation.  I hope this worked out reasonably.  I left a somewhat long "Otherwise" clause at the end of the switch, using existing spec text, because splitting it up felt a bit odd, although not splitting it up also feels a bit odd.
* The suggested `<span data-x="concept-tree-root">root</span>` didn't work, but `<span>root</span>` works fine.

annevk commented on source:
```diff
@@ -13274,8 +13274,18 @@ Transport Protocol">HTTP&lt;/abbr> today.&lt;/p></code></pre> <!-- DO NOT REWRAP
 
-  <p>If these attributes are omitted from an element, then the language of this element is the same
-  as the language of its parent element, if any.</p>
+  <p>If these attributes are omitted from an element, then the language of this element is:</p>
+
+  <dl class="switch">
+   <dt>If the element's parent is a <span>shadow root</span></dt>
+   <dt>If the element is a <code>slot</code> element whose whose <span>root</span> is a <span>shadow
+   root</span></dt>
+   <dd>the same as the language of that <span>shadow root</span>'s <span
+   data-x="dom-hyperlink-host">host</span>, if any.</dd>
+
+   <dt>Otherwise</dt>
+   <dd>the same as the language of its parent element, if any.</dd>
+  </dl>
```
Do we need this part? I wonder who this is for. It omits some aspects of the normative algorithm and as-is I think it might lead to confusion where folks think they need to implement this.

dbaron replied:
That's reasonable.  I've reduced this change (in the edit I'll push shortly) to just noting the exception for the one case that makes the statement untrue (`<slot>` elements), and in a way that I think will at least show an informed reader that this isn't the complete set of rules.

annevk commented on source:
```diff
@@ -13306,28 +13316,35 @@ Transport Protocol">HTTP&lt;/abbr> today.&lt;/p></code></pre> <!-- DO NOT REWRAP
+   <dt>If the node's parent is a <span>shadow root</span></dt>
+   <dt>If the node is a <code>slot</code> element whose whose <span>root</span> is a <span>shadow
+   root</span></dt>
+   <dd>Use the language of that <span>shadow root</span>'s <span
```
xref language

annevk commented on source:
```diff
@@ -13306,28 +13316,35 @@ Transport Protocol">HTTP&lt;/abbr> today.&lt;/p></code></pre> <!-- DO NOT REWRAP
+   <dt>If the node is a <code>slot</code> element whose whose <span>root</span> is a <span>shadow
+   root</span></dt>
+   <dd>Use the language of that <span>shadow root</span>'s <span
+   data-x="dom-hyperlink-host">host</span>.</dd>
```
This is the wrong xref for host.

annevk commented on source:
```diff
@@ -13306,28 +13316,35 @@ Transport Protocol">HTTP&lt;/abbr> today.&lt;/p></code></pre> <!-- DO NOT REWRAP
+   <dd>Use the language of that <span>shadow root</span>'s <span
+   data-x="dom-hyperlink-host">host</span>.</dd>
+
+   <dt>If the node has an ancestor that is an element</dt>
```
I think we can use "parent element" here. If a node's parent is not an element, it won't have an ancestor that's an element either.

> `If the node has a non-null <span>parent element</span>`

annevk commented on source:
```diff
@@ -13306,28 +13316,35 @@ Transport Protocol">HTTP&lt;/abbr> today.&lt;/p></code></pre> <!-- DO NOT REWRAP
+   <dd><p>Use the <span>language</span> of the nearest ancestor element.</p></dd>
+
+   <dt>Otherwise</dt>
+   <dd><p>If there is a <span>pragma-set default language</span> set, then that is the language of
+   the node. If there is no <span>pragma-set default language</span> set, then language information
+   from a higher-level protocol (such as HTTP), if any, must be used as the final fallback language
+   instead. In the absence of any such language information, and in cases where the higher-level
+   protocol reports multiple languages, the language of the node is unknown, and the corresponding
+   language tag is the empty string.</p></dd>
```
Could you file a follow-up issue for this? I wonder if this fallback makes sense in all cases. In particular for synthetic documents or an XHR document or some such.

dbaron replied:
Filed #9807.

annevk commented on source:
```diff
@@ -13638,11 +13717,20 @@ Transport Protocol">HTTP&lt;/abbr> today.&lt;/p></code></pre> <!-- DO NOT REWRAP
   <ol>
-   <li><p>If <var>element</var> has no <span>parent element</span>, then return '<span
-   data-x="concept-ltr">ltr</span>'.</p></li>
+   <li><p>If <var>element</var> is a <code>slot</code> element and is in a <span>shadow tree</span>,
```
shadow tree -> root is shadow root

annevk commented on source:
```diff
@@ -13586,47 +13613,99 @@ Transport Protocol">HTTP&lt;/abbr> today.&lt;/p></code></pre> <!-- DO NOT REWRAP
-     <ref>BIDI</ref></p></li>
+     <li>
+      <p>If <var>descendant</var> is a <code>slot</code> element that is in a <span>shadow
+      tree</span>, then:</p>
```
root is a shadow root

annevk reviewed (commented):
Thanks, couple more nits and suggestions.

annevk commented on source:
```diff
@@ -13543,7 +13552,17 @@ Transport Protocol">HTTP&lt;/abbr> today.&lt;/p></code></pre> <!-- DO NOT REWRAP
+     <li><p>Let <var>result</var> be the <span>auto directionality</span> of
+     <var>element</var>.</p></li>
+
+     <li><p>If <var>result</var> is null, then set <var>result</var> to the <span>parent
+     directionality</span> of <var>element</var>.</p></li>
```
```suggestion
     <li><p>If <var>result</var> is null, then return the <span>parent directionality</span> of
     <var>element</var>.</p></li>
```

annevk commented on source:
```diff
@@ -13586,47 +13605,99 @@ Transport Protocol">HTTP&lt;/abbr> today.&lt;/p></code></pre> <!-- DO NOT REWRAP
+       <li><p>If <var>child</var> is a <code>Text</code> node, then set <var>childDirection</var> to
+       the <span>text node directionality</span> of <var>child</var>.</p></li>
+
+       <li><p>If <var>child</var> is an <code>Element</code> node, then set
+       <var>childDirection</var> to the <span>auto directionality</span> of
+       <var>child</var>.</p></li>
```
Reading this rendered I think it would make sense to make this more verbose:

> 1. Otherwise:
>    1. Assert: child is an Element node.
>    2. Set ...

That makes it clear there's only two possible branches.

smaug---- reviewed (approved):
Thanks, this looks easy to read and understand.

annevk commented:
Remaining:

* https://github.com/web-platform-tests/wpt/pull/29820 will get some further fixes still. That PR should probably land together with this PR.
* It probably makes sense to file an MDN issue for this?

dbaron commented:
I think web-platform-tests/wpt#29820 should now be sufficiently updated, and part of it is now split into web-platform-tests/wpt#42364.

I'm a little unsure about filing an MDN issue.  Right now neither https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/dir nor https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/lang goes into sufficient detail that they'd be affected by any of the changes here.  That said, maybe it's possible that they should *mention* inheritance, although I'm not sure they should go into anything at the level of detail that would be changed by the changes here.

annevk commented:
They should probably say something either there or in some shadow tree article.
